### DISCUSSION TITLE
getting `numba_mpi.scatter` to work (type of first argument, ...)

### DISCUSSION THREAD

**Discussion by pnvdp** [atmos-cloud-sim-uj/numba-mpi#118] [Q&A]
I want to scatter an numpy array from root = 0 to all processors by numba_mpi.scatter(). Here is the code:
import numba_mpi
import numpy as np
from numba import jit
@jit
def roll_array_scalar():   
    if numba_mpi.rank() == 0:
	    A = np.empty((2,4));
	    A[0] = np.random.uniform(1,1,(4));
	    A[1] = np.random.uniform(2,2,(4));
    else:
        A = None
    B = np.empty((1,4));
    numba_mpi.scatter( A, B, count = B.size, root = 0 )
    print(numba_mpi.rank(),B)
 
roll_array_scalar()

**Here is the error:** numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Unknown attribute 'flags' of type none
File "../../../anaconda3/lib/python3.9/site-packages/numba_mpi/api/scatter_gather.py", line 35:
def scatter(send_data, recv_data, count, root):
    <source elided>
    """wrapper for MPI_Scatter(). Returns integer status code (0 == MPI_SUCCESS)"""
    assert send_data.flags.c_contiguous  # TODO #60
    ^
During: typing of get attribute at /home/nandita/anaconda3/lib/python3.9/site-packages/numba_mpi/api/scatter_gather.py (35)

**However, if I define the array A in all processors and then scatter it to B, then it works:** 
@jit
def roll_array_scalar():   
    A = np.empty((2,4));
    A[0] = np.random.uniform(1,1,(4));
    A[1] = np.random.uniform(2,2,(4));
    B = np.empty((1,4));
    numba_mpi.scatter( A, B, count = B.size, root = numba_mpi.rank() )
    print(numba_mpi.rank(),B)
 
roll_array_scalar()
Output: 0 [[1. 1. 1. 1.]]
1 [[2. 2. 2. 2.]]

Please resolve the issue so that I can scatter an array defined in rank 0 to all processors.

**Comment by david-zwicker** [comment]
I don't fully understand the problem, partly because your code is not indented correctly. Could you please post a working example of your problem formatted as `code`?

**Comment by david-zwicker** [comment]
You need to enclose your code with three backticks on both sides (or use the button in above the comment box) to format code correctly. Use the "Preview" pane to check your post before sending it.

**Comment by pnvdp** [comment]
```python
import numba_mpi
from numba import jit
import numpy as np
#### run this code using 'mpirun -np 2 python file_name.py' #################

@jit
def scatter_from_rank_0_1():   
    A = np.random.uniform(-1,1,(2,4));  # here A is defined in all arrays 
    print("rank and A,", numba_mpi.rank(),A)
    B = np.empty((1,4));
    numba_mpi.scatter( A, B, count = B.size, root = 0) # The Array A (defined in rank 0) is scattered in B ( defined in rank 0 and rank 1)
    print("rank and B,",numba_mpi.rank(),B)
 
scatter_from_rank_0_1()
## ******** scatter_from_rank_0_1 is working ********** ##
## Here A is defined in all arrays and then the A in rank 0 is scattered to B. 
## But I don't want to define A in all processors because I only need to scatter A (in rank 0) to be scattered. Also this will create memmory issue. ###########

@jit
def scatter_from_rank_0_2():   
    if numba_mpi.rank() == 0:
        A = np.random.uniform(-1,1,(2,4));
        print("rank and A,", numba_mpi.rank(),A)
    else:
        A = None
    B = np.empty((1,4));
    numba_mpi.scatter( A, B, count = B.size, root = 0) 
    print(numba_mpi.rank(),B)
 
# uncomment this to run 
# scatter_from_rank_0_2()

## ******** But scatter_from_rank_0_b is not working ********** ##
## Here A is defined in rank 0 and then I want to scatter it to B. The same works well for comm.Scatter for MPI.

## Note that Gather part is working well. #####
@jit
def gather_into_rank_0():   
    B = np.random.uniform(-1,1,(1,4));
    print("rank and B", numba_mpi.rank(), B)
    A = np.empty((2,4))
    numba_mpi.gather( B, A, count = B.size, root = 0 ) #  B ( defined in rank 0 and rank 1) is to be gathered in Array A (defined in rank 0)
    if numba_mpi.rank() == 0:
        print(A)

gather_into_rank_0()

```

**Comment by slayoo** [comment]
The code above uses `scatter` with first argument set to `None` whenever rank is not equal 0, hence the assertion stops execution.

In numba-mpi tests folder, there is a seemingly analogous example where the argument is instead set to an empty 0-element array:
https://github.com/numba-mpi/numba-mpi/blob/9750d2878bca5627503c7dbc0a2a383fc6a81ffd/tests/api/test_scatter_gather.py#L39

Would that help?

**Comment by pnvdp** [comment]
Thank you for the suggestion. I have  implemented the above , but still getting some error:
```python
@jit
def scatter_from_rank_0_2():   
    if numba_mpi.rank() == 0:
        A = np.random.uniform(-1,1,(2,4,4));
        print("rank and A,", numba_mpi.rank(),A)
    else:
        A = np.empty(shape=(0,), dtype=float) 

    B = np.empty((1,4,4), dtype = float);
    numba_mpi.scatter( A, B, count = B.size, root = 0) 
    print(numba_mpi.rank(),B)
```
Error:Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.

For more information visit https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit

File "scatter_numba_mpi.py", line 21:
@jit
def scatter_from_rank_0_2():

**Comment by slayoo** [comment]
> Error:Fall-back from the nopython compilation path to the object mode compilation path has been detected, this is deprecated behaviour.

Please try using the `@njit` decorator instead.

**Comment by pnvdp** [comment]
```python
@njit
def scatter_from_rank_0_2():   
    if numba_mpi.rank() == 0:
        A = np.random.uniform(-1,1,(2,4,4));
        print("rank and A,", numba_mpi.rank(),A)
    else:
        A = np.empty(shape=(0,), dtype=float)

    B = np.empty((1,4,4), dtype=float);
    numba_mpi.scatter( A, B, count = B.size, root = 0) 
    print(numba_mpi.rank(),B)
```
 
I got following error: No implementation of function Function(<built-in function empty>) found for signature:
  empty(shape=UniTuple(Literal[int](0) x 1), dtype=Function(<class 'float'>))
 
There are 4 candidate implementations:
   - Of which 2 did not match due to:
   Overload in function '_OverloadWrapper._build.<locals>.ol_generated': File: numba/core/overload_glue.py: Line 131.
     With argument(s): '(shape=UniTuple(int64 x 1), dtype=Function(<class 'float'>))':
    Rejected as the implementation raised a specific error:
      TypingError: Failed in nopython mode pipeline (step: nopython frontend)
    No implementation of function Function(<intrinsic stub>) found for signature:
     stub(UniTuple(int64 x 1), Function(<class 'float'>))
      There are 2 candidate implementations:
      - Of which 2 did not match due to:
      Intrinsic of function 'stub': File: numba/core/overload_glue.py: Line 35.
        With argument(s): '(UniTuple(int64 x 1), Function(<class 'float'>))':
       No match.
    During: resolving callee type: Function(<intrinsic stub>)
    During: typing of call at <string> (3)
    <source missing, REPL/exec in use?>

  raised from /home/nandita/anaconda3/lib/python3.9/site-packages/numba/core/typeinfer.py:1086
   - Of which 2 did not match due to:
   Overload in function '_OverloadWrapper._build.<locals>.ol_generated': File: numba/core/overload_glue.py: Line 131.
     With argument(s): '(shape=UniTuple(Literal[int](0) x 1), dtype=Function(<class 'float'>))':
    Rejected as the implementation raised a specific error:
      TypingError: Failed in nopython mode pipeline (step: nopython frontend)
    No implementation of function Function(<intrinsic stub>) found for signature:
      stub(UniTuple(Literal[int](0) x 1), Function(<class 'float'>))
      There are 2 candidate implementations:
      - Of which 2 did not match due to:
      Intrinsic of function 'stub': File: numba/core/overload_glue.py: Line 35.
        With argument(s): '(UniTuple(int64 x 1), Function(<class 'float'>))':
       No match.
    During: resolving callee type: Function(<intrinsic stub>)
    During: typing of call at <string> (3)
  File "<string>", line 3:
    <source missing, REPL/exec in use?> raised from /home/nandita/anaconda3/lib/python3.9/sitepackages/numba/core/typeinfer.py:1086
During: resolving callee type: Function(<built-in function empty>)
During: typing of call at /home/nandita/Downloads/HIGH PERFORMANCE COMPUTING/My_nandita_code/scatter_numba_mpi.py (26)
File "scatter_numba_mpi.py", line 26:
def scatter_from_rank_0_2():   
    <source elided>
    else:
        A = np.empty(shape=(0,), dtype=float)
        ^

**Comment by slayoo** [comment]
This could be related to an apparently long-standing issue within Numba: https://github.com/numba/numba/issues/3068#issuecomment-1014424145
Please try `dtype=np.float64` instead of `dtype=float`

**Reply by pnvdp** [reply]
Implementing np.float64
```@njit
def scatter_from_rank_0_2():   
    if numba_mpi.rank() == 0:
        A = np.random.uniform(-1,1,(2,4,4));
        print("rank and A,", numba_mpi.rank(),A)
    else:
        A = np.empty(shape=(0,), dtype=np.float64)

    B = np.empty((1,4,4), dtype=np.float64);
    numba_mpi.scatter( A, B, count = B.size, root = 0) 
    print(numba_mpi.rank(),B)
```
I am getting following error:
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Cannot unify array(float64, 1d, C) and array(float64, 3d, C) for 'A.2', defined at /home/nandita/Downloads/HIGH PERFORMANCE COMPUTING/My_nandita_code/scatter_numba_mpi.py (29)
File "scatter_numba_mpi.py", line 29:
def scatter_from_rank_0_2():   
    <source elided>
B = np.empty((1,4,4), dtype=np.float64);
    ^During: typing of assignment at /home/nandita/Downloads/HIGH PERFORMANCE COMPUTING/My_nandita_code/scatter_numba_mpi.py (29)
